pkg/sign/notary.go

```diff
@@ -36,214 +40,357 @@ func (e ErrBadResponse) Error() string {
-		// wait for ticker to run
-		<-w.C
-		resp, err := ns.c.Do(req)
-		if err != nil {
-			return fmt.Errorf("notary request: %w", err)
+	if resp.StatusCode != http.StatusAccepted {
+		respMsg, _ := io.ReadAll(resp.Body)
+		return ErrBadResponse{
+			status:  resp.Status,
+			message: fmt.Sprintf("failed to sign images: %s", string(respMsg)),
 		}
-		status = resp.Status
-		respMsg, err = io.ReadAll(resp.Body)
+	}
+
+	fmt.Printf("Successfully signed images %s!\n", sImg)
+	return nil
+}
+
+func retryHTTPRequest(client *http.Client, req *http.Request, retries int, retryInterval time.Duration) (*http.Response, error) {
+	var lastResp *http.Response
+	var lastErr error
+
+	// Read and store the request body
+	var bodyBytes []byte
+	if req.Body != nil {
+		var err error
+		bodyBytes, err = io.ReadAll(req.Body)
 		if err != nil {
-			return fmt.Errorf("body read: %w", err)
+			return nil, ErrBadResponse{status: "400", message: fmt.Sprintf("failed to read request body: %v", err)}
 		}
-		switch resp.StatusCode {
-		case http.StatusOK:
-			// response was fine. Do not need anything else
-			fmt.Printf("Successfully signed images %s!\n", sImg)
-			return nil
-		case http.StatusUnauthorized, http.StatusForbidden, http.StatusBadRequest, http.StatusUnsupportedMediaType:
-			return ErrBadResponse{status: status, message: string(respMsg)}
+		req.Body.Close()
+	}
+
+	for retries > 0 {
+		// Reset the request body for each retry
+		if bodyBytes != nil {
+			req.Body = io.NopCloser(bytes.NewReader(bodyBytes))
 		}
+
+		resp, err := client.Do(req)
+		if err != nil {
+			lastErr = err
+		} else if resp.StatusCode == http.StatusAccepted {
+			return resp, nil
+		} else {
+			lastErr = fmt.Errorf("failed to sign images, unexpected status code: %d", resp.StatusCode)
+		}
+
+		lastResp = resp
 		retries--
+		if retries == 0 {
+			break
+		}
+		time.Sleep(retryInterval)
+	}
+	return lastResp, ErrBadResponse{status: "500", message: fmt.Sprintf("request failed after retries: %v", lastErr)}
+}
+
+func setupTLS(cert tls.Certificate) *tls.Config {
+	return &tls.Config{
+		Certificates: []tls.Certificate{cert},
 	}
-	fmt.Println("Reached all retries. Stopping.")
-	return ErrBadResponse{status: status, message: string(respMsg)}
 }
 
 func (nc NotaryConfig) NewSigner() (Signer, error) {
-	var ns NotarySigner
+	ns := NotarySigner{
+		retryTimeout:       nc.RetryTimeout,
+		url:                nc.Endpoint,
+		ParseReferenceFunc: ParseReference,
+		GetImageFunc:       GetImage,
+		SetupTLSFunc:       setupTLS,
+	}
+
+	ns.BuildPayloadFunc = ns.buildPayload
+	ns.DecodeCertFunc = ns.signifySecret.DecodeCertAndKey
+
+	// HTTP client configuration
+	ns.c = &http.Client{
+		Timeout: nc.Timeout,
+	}
 
-	// (@Ressetkk): Should this be loaded before calling signer?
 	if nc.Secret != nil {
-		f, err := os.ReadFile(nc.Secret.Path)
-		if err != nil {
-			return nil, err
-		}
 		switch nc.Secret.Type {
-		case "token":
-			ns.authFunc = AuthToken(string(f))
 		case "signify":
```

Comment:
For last two years we didn't have other signers. We have no any plans to support more than one signer. I would focus on building proper abstraction on signer implementation to be able easily change it, rather than having support for multiple signers.

Reply:
Removed this.